### PR TITLE
StaticLLMPipeline: Support more generation options

### PR DESCRIPTION
**Pull request by TolyaTalamanov**

### CODE REVIEW

**Comment by ilya-lavrenov** [comment]
why not to re-use the whole Sampler? Limited to cases when we don't need to fork sequences (like in beam search)

**Comment by TolyaTalamanov** [comment]
> why not to re-use the whole Sampler? Limited to cases when we don't need to fork sequences (like in beam search)

I was a little bit confused about the API contract for `SequenceGroup`, this is what I've got from stateful pipeline implementaiton:
```
auto sequence_group = std::make_shared<SequenceGroup>(request_id, tokenized_chat_history, config, block_size, enable_prefix_caching);
auto sequences_group = std::vector{sequence_group};

// (1) Why?
for (sequence_group : sequences_group) {
    sequence_group->set_sequence_group_ptr(sequence_group);
}

auto active_sequence_group{sequence_groups};

// Prefill
...
infer()
for (auto sequence_group : sequences_group) {
    int64_t sequence_len = logits.get_shape().at(1);
    sequence_group->update_processed_tokens_num(sequence_group->get_prompt_len() - sequence_len);
    sequence_group->schedule_tokens(sequence_len);
}

while (!active_sequence_group) {
    // Generation

    // (2) Why not active_sequence_group?
    for (auto sequence_group : sequences_group) {
        sequence_group->schedule_tokens(1);
        size_t num_sequences = sequence_group->num_running_seqs();
        total_num_tokens += sequence_group->get_num_scheduled_tokens() * num_sequences;
    }

    for (auto seqence_group : active_sequence_group) {
         std::vector<Sequence::Ptr> running_sequences = sequence_group->get_running_sequences()
         size_t num_running_sequences = running_sequences.size();
         size_t num_scheduled_tokens = sequence_group->get_num_scheduled_tokens();
         size_t group_position_id = sequence_group->get_num_processed_tokens();
         // (3) Not needed for multinomial / greedy?
         std::map<size_t, int32_t> beam_idxs = sampler.get_beam_idxs(sequence_group);
         // (4) Can it be also skipped for greedy / multinomial?
         for (size_t seq_id = 0; seq_id < num_running_sequences; ++seq_id) {
             Sequence::CPtr sequence = running_sequences[seq_id];

             for (size_t token_id = 0, position_id = group_position_id; token_id < num_scheduled_tokens; ++token_id, ++position_id) {
                 // compute token for current sequence
                 input_ids_data[token_id] = position_id < sequence_group->get_prompt_len() ?
                 sequence_group->get_prompt_ids()[position_id] :
                 sequence->get_generated_ids()[position_id - sequence_group->get_prompt_len()];
                }

                // apply strides to shift to a next sequence
                input_ids_data += num_scheduled_tokens;

                // for different sequences iteration of beams started from 0, but we collect it to one input_ids
                next_beams.push_back(beam_idxs[sequence->get_id()] + beam_offets.at(sequence_group->get_request_id()));
         }
         ...
         infer()
         sampler_output = sampler.sample(active_sequence_groups, m_llm.get_tensor("logits"));
         // push tokens to streamer
    }
    // push tokens from every sequence to the result
}
```

But I've got your point, maybe "heavy" part can be dropped, as we don't need `beam_search`.

@ilya-lavrenov Could you double check if my understanding is correct of how `SequenceGroup` API should be used?
```
auto sequence_group = std::make_shared<SequenceGroup>(request_id, tokenized_chat_history, config, block_size, enable_prefix_caching);
sequence_group->set_sequence_group_ptr(sequence_group);

// Prefill
infer();
sequence_group->update_processed_tokens_num(sequence_group->get_prompt_len() - sequence_len);
sequence_group->schedule_tokens(sequence_len);
SamplerOutput sampler_output = sampler.sample({sequence_group}, logits);
// ... Pull tokens

// Generation
for (;;) {
    infer();
    sequence_group->schedule_tokens(1);
    SamplerOutput sampler_output = sampler.sample({sequence_group}, logits);
    // ... Pull tokens
}

// Must be always 1?
const auto& sequence = sequence_group->get_finished_sequences().front();
results.tokens.push_back(sequence->get_generated_ids());
results.scores.push_back(sequence->get_cumulative_log_probs());
```

**Comment by ilya-lavrenov** [comment]
Regarding your inline questions:

(1) https://github.com/openvinotoolkit/openvino.genai/pull/1456
(2) It's already updated to use `active_sequence_groups`, see https://github.com/openvinotoolkit/openvino.genai/blob/0c5f03ba05e4332476398108f3681ae865e5d3a1/src/cpp/src/lm_encoding.cpp#L137-L145
(3) No, `beam_idx` is required only for beam search and parallel sampling (when we generate multiple answers for the same prompt using `do_sample=True`). BTW, `beam_idx ` is not supported by static LLM pipeline, right? So, looks like it's not required anyway
(4) Multiple sequences (concurrently generated answers) within the same sequence group (shares prompt between sequences, etc) are required only for beam search / parallel sampling.

IMO, your simplified code snippet looks correct and serves well for static pipeline case.

File: src/cpp/src/llm_pipeline_static.cpp [hunk 1]
```diff
@@ -235,12 +238,12 @@ enum class GenerateHint {
 
 std::string to_string(GenerateHint h) {
     switch(h) {
-        case GenerateHint::FAST_COMPILE :
```
Comment (TolyaTalamanov) [inline src/cpp/src/llm_pipeline_static.cpp:238]: It really requires linter...

File: tests/python_tests/test_llm_pipeline_static.py [hunk 2]
```diff
@@ -29,25 +51,66 @@ def generate_chat_history(model_path, device, pipeline_config, questions):
...
+    get_multinomial_temperature_and_repetition_penalty(),
+    get_multinomial_temperature_and_frequence_penalty(),
+    get_multinomial_temperature_and_presence_penalty()
+]
+@pytest.mark.skipif(sys.platform in ["darwin", "linux"], reason="Not supposed to work on mac. Segfault on linux CI")
+@pytest.mark.precommit
+@pytest.mark.nightly
+@pytest.mark.parametrize("generation_config", generation_configs)
+def test_multinomial_sampling(generation_config):
+	# Multinomial sampling is highly sensitive to raw logits values. For fair comparison,
```
[the remark this answers is not in the document]
Comment (TolyaTalamanov) [reply tests/python_tests/test_llm_pipeline_static.py]: Done

File: src/cpp/src/sampler.cpp [hunk 3]
```diff
@@ -67,6 +67,71 @@ std::vector<Token> log_softmax(const ov::Tensor& logits, size_t batch_idx) {
     return tokens;
 }
 
+Token greedy_sample(const Logits& logits, size_t top_logprobs) {
```
[the remark this answers is not in the document]
Comment (TolyaTalamanov) [reply src/cpp/src/sampler.cpp]: Removed

File: src/cpp/src/llm_pipeline_static.hpp [hunk 4]
```diff
@@ -4,8 +4,10 @@
 #pragma once
 
 #include <filesystem>
+#include <random>
```
[the remark this answers is not in the document]
Comment (TolyaTalamanov) [reply src/cpp/src/llm_pipeline_static.hpp]: Removed

File: src/cpp/src/llm_pipeline_static.cpp [hunk 5]
```diff
@@ -938,6 +945,31 @@ DecodedResults StaticLLMPipeline::generate(
     return decoded_results;
 }
 
+void stream_generated_tokens(std::shared_ptr<StreamerBase> streamer_ptr,
```
[the remark this answers is not in the document]
Comment (TolyaTalamanov) [reply src/cpp/src/llm_pipeline_static.cpp]: Done

File: src/cpp/src/llm_pipeline_static.cpp [hunk 6]
```diff
@@ -938,6 +945,31 @@ DecodedResults StaticLLMPipeline::generate(
...
+                             GenerationHandle& handle) {
+    if (streamer_ptr && handle->can_read()) {
+        std::unordered_map<uint64_t, GenerationOutput> token = handle->back();
+        for (const auto& gen_token : token.begin()->second.generated_ids) {
+            if (streamer_ptr->put(gen_token)) {
+                handle->drop();
```
[the remark this answers is not in the document]
Comment (TolyaTalamanov) [reply src/cpp/src/llm_pipeline_static.cpp]: Added

File: src/cpp/src/llm_pipeline_static.cpp [hunk 7]
```diff
@@ -1016,11 +1048,21 @@ EncodedResults StaticLLMPipeline::generate(
...
-        return results;
-    }
+
+    auto logits = m_prefill_request.get_tensor("logits");
+    int64_t output_sequence_len = logits.get_shape().at(1);
+
+    auto sequence_group = std::make_shared<SequenceGroup>(
+        0 /* request_id */, padded_input_ids, config, 1 /* block_size */);
+    sequence_group->update_processed_tokens_num(m_kvcache_desc.max_prompt_size - output_sequence_len);
+    sequence_group->schedule_tokens(output_sequence_len);
```
Comment (ilya-lavrenov) [inline src/cpp/src/llm_pipeline_static.cpp:1058]: here values passed to `update_processed_tokens_num` and `schedule_tokens` should be in total give `prompt_len`

File: src/cpp/src/llm_pipeline_static.cpp [hunk 8]
```diff
@@ -632,6 +634,31 @@ void copy_columns_by_row_chunks(const ov::Tensor& src, ov::Tensor& dst) {
...
+int64_t get_last_token(ov::genai::SequenceGroup::Ptr sequence_group) {
+    const auto running_sequences = sequence_group->get_running_sequences();
+    OPENVINO_ASSERT(running_sequences.size() == 1u);
+    const auto sequence = running_sequences.front();
+
+    size_t num_scheduled_tokens = sequence_group->get_num_scheduled_tokens();
+    OPENVINO_ASSERT(num_scheduled_tokens == 1u);
+
+    const auto num_processed_tokens = sequence_group->get_num_processed_tokens();
+    return sequence->get_generated_ids()[num_processed_tokens - sequence_group->get_prompt_len()];
```
Comment (ilya-lavrenov) [inline src/cpp/src/llm_pipeline_static.cpp:659]: simply `sequence->get_generated_ids().back()` ?

File: src/cpp/src/llm_pipeline_static.cpp (hunk 7 above)
Comment (TolyaTalamanov) [reply src/cpp/src/llm_pipeline_static.cpp:1058]: `prompt_len` in case of static pipeline is `max_prompt_size`, so it must be the correct setting

File: src/cpp/src/llm_pipeline_static.cpp (hunk 8 above)
Comment (TolyaTalamanov) [reply src/cpp/src/llm_pipeline_static.cpp]: Done

File: tests/python_tests/test_llm_pipeline_static.py [hunk 9]
```diff
@@ -101,16 +164,17 @@ def test_batch_raise_error():
...
 @pytest.mark.parametrize("generation_config", generation_configs)
 @pytest.mark.precommit
 @pytest.mark.nightly
 def test_unsupported_sampling_raise_error(generation_config):
     model_path = get_models_list()[0][1]
-    prompt = 'The Sun is yellow because'
+    prompt = 'What is OpenVINO?'
     pipe = ov_genai.LLMPipeline(model_path, "NPU", **common_config)
     with pytest.raises(RuntimeError):
         pipe.generate(prompt, **generation_config)
```
Comment (ilya-lavrenov) [inline tests/python_tests/test_llm_pipeline_static.py:180]:
```suggestion
        pipe.generate(prompt, generation_config)
```

File: tools/llm_bench/task/text_generation.py [hunk 10]
```diff
@@ -227,7 +227,6 @@ def run_text_generation_genai(input_text, num, model, tokenizer, args, iter_data
     gen_config = model.get_generation_config()
     gen_config.max_new_tokens = max_gen_tokens
     gen_config.num_beams = args["num_beams"]
```
Comment (sbalandi) [inline tools/llm_bench/task/text_generation.py:229]: please, remove `set_seed(args['seed'])` on str 201 and 356 and add `gen_config.rng_seed= args["seed"]` and to here `run_text_generation_genai_with_stream`

Comment (TolyaTalamanov) [reply tools/llm_bench/task/text_generation.py:229]: Done, thanks!

**Review by sbalandi — APPROVED** [review]

File: tests/python_tests/test_llm_pipeline_static.py [hunk 11]
```diff
@@ -29,25 +51,66 @@ def generate_chat_history(model_path, device, pipeline_config, questions):
...
+    get_greedy_with_repetition_penalty(),
+    get_greedy_with_penalties(),
+    get_greedy_with_min_and_max_tokens(),
+    get_greedy_with_single_stop_string(),
+    get_greedy_with_multiple_stop_strings(),
+    get_greedy_with_multiple_stop_strings_no_match(),
+    get_greedy_stop_strings_exclude_from_output(),
+    get_greedy_stop_strings_include_to_output(),
+    get_greedy_n_stop_strings_exclude_from_output(),
+    get_greedy_n_stop_strings_include_to_output()
```
Comment (ilya-lavrenov) [inline tests/python_tests/test_llm_pipeline_static.py:66]: do we really need all these variants? here you are rather testing Sampler than static pipeline.

Sampler itself is tested in test_sampling.py 

I suppose you can keep greedy and something with stop strings (to ensure that GenerationHandle is used instead of manual reading from SequenceGroup)

File: tests/python_tests/test_llm_pipeline_static.py [hunk 12]
```diff
@@ -29,25 +51,66 @@ def generate_chat_history(model_path, device, pipeline_config, questions):
...
+    # Multinomial sampling is highly sensitive to raw logits values. For fair comparison,
+    # a reference implementation producing identical logits (e.g., from StaticLLMPipeline)
+    # would be necessary. However, the CPU in StatefulPipeline and StaticLLMPipeline may apply
+    # different optimizations due to differences in provided topologies, leading to slight
+    # variations in raw logits. Therefore, there is no reliable reference for validation,
+    # so only ensure that no exceptions are raised.
+    prompt = 'What is OpenVINO?'
+    model_path = get_models_list()[0][1]
+    static_pipe = ov_genai.LLMPipeline(model_path, "NPU", **common_config)
+    actual_out = static_pipe.generate(prompt, generation_config)
```
Comment (ilya-lavrenov) [inline tests/python_tests/test_llm_pipeline_static.py:111]: you don't even compare the results - do we need all these tests?

To test that Sampler is integrated correctly, you don't need to iterate over all combination of sampling parameters.

Comment (TolyaTalamanov) [reply tests/python_tests/test_llm_pipeline_static.py:80]: Let it be at least one run, to check if `multinomial` doesn't throw

File: tools/llm_bench/task/text_generation.py [hunk 13]
```diff
@@ -379,9 +377,9 @@ def run_text_generation_genai_with_stream(input_text, num, model, tokenizer, arg
     max_gen_tokens = DEFAULT_OUTPUT_TOKEN_SIZE if args['infer_count'] is None else args['infer_count']
     streamer.reset()
     gen_config = model.get_generation_config()
+    gen_config.rng_seed = args["seed"]
     gen_config.max_new_tokens = max_gen_tokens
     gen_config.num_beams = args["num_beams"]
-    gen_config.do_sample = False
```
Comment (ilya-lavrenov) [inline tools/llm_bench/task/text_generation.py:384]: we need to revert it once GenerationConfig::validate is fixed, otherwise OV validation will validate multinomial instead of greedy after this change.

Comment (TolyaTalamanov) [reply tools/llm_bench/task/text_generation.py:384]: But now, some of the models failed because their config have `top_p != 1.0`  but `llm_bench` forces `do_sample=False`

Comment (ilya-lavrenov) [reply tools/llm_bench/task/text_generation.py:384]: PR https://github.com/openvinotoolkit/openvino.genai/pull/1473

Comment (TolyaTalamanov) [reply tools/llm_bench/task/text_generation.py:384]: Returned back

File: tests/python_tests/test_llm_pipeline_static.py [hunk 14]
```diff
@@ -29,25 +36,50 @@ def generate_chat_history(model_path, device, pipeline_config, questions):
     return chat_history
 
 
+generation_configs = [
+    get_greedy_n_stop_strings_exclude_from_output(),
+    get_greedy_n_stop_strings_include_to_output()
+]
 @pytest.mark.skipif(sys.platform in ["darwin", "linux"], reason="Not supposed to work on mac. Segfault on linux CI")
```
Comment (ilya-lavrenov) [inline tests/python_tests/test_llm_pipeline_static.py:43]: please, disable all test on arm64 as well.

BTW, you can skip all tests on module level https://docs.pytest.org/en/stable/how-to/skipping.html#id1 where NPU plugin is not available

File: tests/python_tests/test_llm_pipeline_static.py [hunk 15]
```diff
@@ -132,6 +165,49 @@ def test_max_number_of_tokens():
...
+    pipe = ov_genai.LLMPipeline(model_path, "NPU", **pipeline_config)
+    encoded_results = pipe.generate(tokenized_input, max_new_tokens=1000, ignore_eos=True)
+
+    assert len(encoded_results.tokens[0]) == (kv_cache_size - input_len + 1)
+
+
+@pytest.mark.skipif(sys.platform in ["darwin", "linux"], reason="Not supposed to work on mac. Segfault on linux CI")
+@pytest.mark.precommit
+@pytest.mark.nightly
+def test_terminate_by_sampler():
```
Comment (ilya-lavrenov) [inline tests/python_tests/test_llm_pipeline_static.py:191]:
```suggestion
def test_terminate_by_callback():
```
